In [1]:
import json
import pprint

In [32]:
i = 0
with open('yelp_dataset/yelp_academic_dataset_tip.json') as tipfile:
    for line in tipfile:
        json_data = json.loads(line.strip())
        json_data

        i += 1
#         print(line)
        if i == 3:
            break

In [ ]:
print(json_data)

In [10]:
tipfile = open('yelp_dataset/yelp_academic_dataset_tip.json')
tipdata = tipfile.readline()

In [11]:
tipdata

'{"user_id": "ZxdojCZgZhw36xuSGaryRg", "text": "Pizza is garbage. Hoagies are excellent.", "business_id": "5UmKMjUEUNdYWqANhGckJw", "likes": 0, "date": "2016-07-17", "type": "tip"}\n'

In [12]:
tipdata.strip()

'{"user_id": "ZxdojCZgZhw36xuSGaryRg", "text": "Pizza is garbage. Hoagies are excellent.", "business_id": "5UmKMjUEUNdYWqANhGckJw", "likes": 0, "date": "2016-07-17", "type": "tip"}'

In [8]:
json.loads('{"favorited": false, "contributors": null}')

{'contributors': None, 'favorited': False}

In [19]:
json.loads('{"user_id": "ZxdojCZgZhw36xuSGaryRg", "text": "Pizza is garbage. Hoagies are excellent.", "business_id": "5UmKMjUEUNdYWqANhGckJw", "likes": 0, "date": "2016-07-17", "type": "tip"}')

{'business_id': '5UmKMjUEUNdYWqANhGckJw',
 'date': '2016-07-17',
 'likes': 0,
 'text': 'Pizza is garbage. Hoagies are excellent.',
 'type': 'tip',
 'user_id': 'ZxdojCZgZhw36xuSGaryRg'}

In [21]:
json.loads(tipdata.strip())

{'business_id': '5UmKMjUEUNdYWqANhGckJw',
 'date': '2016-07-17',
 'likes': 0,
 'text': 'Pizza is garbage. Hoagies are excellent.',
 'type': 'tip',
 'user_id': 'ZxdojCZgZhw36xuSGaryRg'}

# Using Pandas read_json function:

In [3]:
import pandas as pd

In [5]:
with open('yelp_dataset/yelp_academic_dataset_tip.json') as file:
        tip = pd.read_json(file, lines = True)

In [6]:
with open('yelp_dataset/yelp_academic_dataset_review.json') as file:
        review = pd.read_json(file, lines = True)

In [7]:
with open('yelp_dataset/yelp_academic_dataset_business.json') as file:
        business = pd.read_json(file, lines = True)

In [609]:
business[business.state == ''].index

Int64Index([], dtype='int64')

In [608]:
#drop one entry with no state:
business.drop(business[business.state == ''].index, inplace=True)

In [8]:
with open('yelp_dataset/yelp_academic_dataset_checkin.json') as file:
        checkin = pd.read_json(file, lines = True)

In [9]:
with open('yelp_dataset/yelp_academic_dataset_user.json') as file:
        user = pd.read_json(file, lines = True)

In [8]:
# dfs = ['tip', 'review', 'business', 'checkin', 'user']

In [9]:
print(tip.columns.values)

['business_id' 'date' 'likes' 'text' 'type' 'user_id']


tip:  ['business_id' 'date' 'likes' 'text' 'type' 'user_id']
review:  ['business_id' 'date' 'review_id' 'stars' 'text' 'type' 'user_id' 'votes']
business:  ['attributes' 'business_id' 'categories' 'city' 'full_address' 'hours'
 'latitude' 'longitude' 'name' 'neighborhoods' 'open' 'review_count'
 'stars' 'state' 'type']
checkin:  ['business_id' 'checkin_info' 'type']
user:  ['average_stars' 'compliments' 'elite' 'fans' 'friends' 'name'
 'review_count' 'type' 'user_id' 'votes' 'yelping_since']

In [10]:
# for df in dfs:
#     print(df+': ', eval(df).columns.values)

In [12]:
user.head(1)

average_stars                                        compliments  \
0           4.14  {'writer': 9, 'more': 3, 'cute': 15, 'cool': 7...   

          elite  fans                                            friends  \
0  [2005, 2006]    69  [rpOyqD_893cqmDAtJLbdog, 4U9kSBLuBDU391x6bxU-Y...   

     name  review_count  type                 user_id  \
0  Russel           108  user  18kPq7GPye-YQ3LyKyAZPw   

                                        votes yelping_since  
0  {'funny': 167, 'cool': 246, 'useful': 282}       2004-10

In [112]:
review[review.user_id=='18kPq7GPye-YQ3LyKyAZPw'] # user: https://www.yelp.com/user_details?userid=c6HT44PKCaXqzN_BdgKPCw
# business: https://www.yelp.com/biz/papa-dels-pizza-champaign-4

business_id       date               review_id  stars  \
539856  zMN8UGd1zDEreT58OCdnyg 2004-10-19  HdODbQJU0D814XdwBxXRfw      5   

                                                     text    type  \
539856  Best pizza I've ever had.  My favorite is the ...  review   

                       user_id                                 votes  
539856  18kPq7GPye-YQ3LyKyAZPw  {'funny': 0, 'cool': 0, 'useful': 2}

In [113]:
review[review.user_id=='18kPq7GPye-YQ3LyKyAZPw'].business_id.values

array(['zMN8UGd1zDEreT58OCdnyg'], dtype=object)

In [39]:
business[business.business_id ==review[review.user_id=='18kPq7GPye-YQ3LyKyAZPw'].business_id.values[0]]

attributes  \
20017  {'Good For Groups': True, 'Has TV': True, 'Cor...   

                  business_id                               categories  \
20017  zMN8UGd1zDEreT58OCdnyg  [Pizza, Sandwiches, Salad, Restaurants]   

            city                         full_address  \
20017  Champaign  1201 S Neil St\nChampaign, IL 61820   

                                                   hours   latitude  \
20017  {'Monday': {'open': '10:30', 'close': '23:00'}...  40.103242   

       longitude              name neighborhoods  open  review_count  stars  \
20017 -88.244164  Papa Del's Pizza            []  True           305    3.5   

      state      type  
20017    IL  business

In [114]:
user[user.user_id == '18kPq7GPye-YQ3LyKyAZPw']

average_stars                                        compliments  \
0           4.14  {'writer': 9, 'more': 3, 'cute': 15, 'cool': 7...   

          elite  fans                                            friends  \
0  [2005, 2006]    69  [rpOyqD_893cqmDAtJLbdog, 4U9kSBLuBDU391x6bxU-Y...   

     name  review_count  type                 user_id  \
0  Russel           108  user  18kPq7GPye-YQ3LyKyAZPw   

                                        votes yelping_since  
0  {'funny': 167, 'cool': 246, 'useful': 282}       2004-10

In [134]:
user[user.user_id == '18kPq7GPye-YQ3LyKyAZPw'].friends[0]

['rpOyqD_893cqmDAtJLbdog',
 '4U9kSBLuBDU391x6bxU-YA',
 'fHtTaujcyKvXglE33Z5yIw',
 '8J4IIYcqBlFch8T90N923A',
 'wy6l_zUo7SN0qrvNRWgySw',
 'HDQixQ-WZEV0LVPJlIGQeQ',
 'T4kuUr_iJiywOPdyM7gTHQ',
 'z_5D4XEIlGAPjG3Os9ix5A',
 'i63u3SdbrLsP4FxiSKP0Zw',
 'pnrGw4ciBXJ6U5QB2m0F5g',
 'ytjCBxosVSqCOQ62c4KAxg',
 'r5uiIxwJ-I-oHBkNY2Ha3Q',
 'niWoSKswEbooJC_M7HMbGw',
 'kwoxiKMyoYjB1wTCYAjYRg',
 '9A8OuP6XwLwnNb9ov3_Ncw',
 '27MmRg8LfbZXNEHkEnKSdA',
 'uguXfIEpI65jSCH5MgUDgA',
 '6VZNGc2h2Bn-uyuEXgOt5g',
 'AZ8CTtwr-4sGM2kZqF6qig',
 'S742m-AuQicMSLDdErrLZQ',
 'uGmQ6ab4iVpWn5m61VFhkQ',
 'GJYJX4SujVj3BR8v2F9PDQ',
 '3shjifK-vZkIHciyy_KbYA',
 '4lc_H2Cf7CO0tCgyA3aSVQ',
 'Tunkp_F1R_uFBJQTsDxD4g',
 'B9pKfr27czBbCoAIircZdQ',
 'pePGMO6EbDpbaZ7D2m6HIg',
 'XRM8W6HUoXbrYKR3BCj9Rg',
 '8DqIWXsKXOipfduYEfFpNw',
 'dvRVX54Z9f7Om51NsTRX1w',
 'CM0saLQmk4oAB17UmQTV-g',
 'HANb8-8InWnju-XzBQQSBw',
 'JuJeZeQJgv7bUreY7a1SlQ',
 '2NVVEEYhhoVELdaPILFrDQ',
 'e4M9_S-ASmRys3DvKQfotw',
 'XA109o963exKgoVGcg9z7w',
 'tVAKdax3Jbf24R7OJB99JQ',
 

In [121]:
user.votes[0]

{'cool': 246, 'funny': 167, 'useful': 282}

In [52]:
review[review.user_id == 'Exg4LdqBrB0dB6jmSsp7AQ']

business_id       date               review_id  stars  \
223231  r1MmMNkS7A0a9UXxXd7qhw 2007-12-28  GM0Ag3P2wtbHt5ZEnEWVDQ      5   
485247  8_HGHlHsNdCnTsLQYHteAw 2008-09-09  XUJcVpI9SQQtzQs35XWntg      5   

                                                     text    type  \
223231  One of the best Mani-Pedi's I've ever had. Ver...  review   
485247  The best Steak on the Strip everything conside...  review   

                       user_id                                 votes  
223231  Exg4LdqBrB0dB6jmSsp7AQ  {'funny': 0, 'cool': 0, 'useful': 1}  
485247  Exg4LdqBrB0dB6jmSsp7AQ  {'funny': 0, 'cool': 1, 'useful': 1}

In [98]:
review.dtypes

business_id            object
date           datetime64[ns]
review_id              object
stars                   int64
text                   object
type                   object
user_id                object
votes                  object
dtype: object

In [107]:
review[(review.date == '2004-10-19') & (review.stars == 5) & (review.text.str.contains('pizza'))]

business_id       date               review_id  stars  \
539856  zMN8UGd1zDEreT58OCdnyg 2004-10-19  HdODbQJU0D814XdwBxXRfw      5   

                                                     text    type  \
539856  Best pizza I've ever had.  My favorite is the ...  review   

                       user_id                                 votes  
539856  18kPq7GPye-YQ3LyKyAZPw  {'funny': 0, 'cool': 0, 'useful': 2}

In [125]:
review[review.text.str.contains('16 inch')]

business_id       date               review_id  stars  \
212      6o3RK6rTcN3nw-j-r2nQmA 2012-08-26  9s0_JxiIK4bjXvwUsjKPLQ      4   
8585     KEzIorUdFjqCuxbRxKPghA 2012-08-26  2Lup8KxR7UIrGmgOF0LqLA      4   
12926    r8jeeS3k_diihMDSu9IjlA 2011-07-23  tDqA9lZRhjiGcF2l_oVLlQ      5   
35016    LwSNsn2k6v_Nb2dbhP3qsQ 2012-06-02  4Ym-J7JHWvNo77xHEO_yww      1   
50849    oCA2OZcd_Jo_ggVmUx3WVw 2011-12-12  tMHcA1xPNgNWd8Kol5jbhA      4   
56445    PEHJjxdw11TGFma9Zhz3ng 2015-06-29  Qew6TVQ_arDSJbGLc84jRg      5   
95016    -KZA7UoULw_pM3jBBUzb8A 2013-07-17  PbHd_7dXVp25Fr6I-ksKDA      2   
165200   LMG0zsAkUSscIvmV9vvm3A 2015-12-16  CUf9TQqymewSweq980Mk7g      1   
170147   m5hMJ7SPIK7are8SykvlvA 2015-03-13  v3vKR-m6aMST_6g7GbpWkw      1   
174879   CdOb17NMdQpagyHSxv5mHA 2010-05-04  uR3hOiWde7xoqySzFmdIlw      4   
176204   9IbQxDmWBG2t8w9nw9VgEA 2008-10-23  oYVdXvYW_o3yz6EsRC89UA      3   
176410   9IbQxDmWBG2t8w9nw9VgEA 2014-10-12  BIKxqhKz9r9o83uh00jw-g      2   
176777   6H_-cKOUkWdoRMS2KrWkNw 2007-03-05  cFMxc4ecBRafFJiwyyNVHw      3   
201879   hr41l2QX7OHh3k_x0Zk2Vw 2010-03-31  bDpBCHSJZ6E9RjbaruMrzA      2   
207061   secsGLdQOaJPATB9SnHlew 2011-07-16  KjNpRf0yAqerXyMMpDCIiQ      5   
207084   secsGLdQOaJPATB9SnHlew 2012-06-01  qQCAZ5VdaqE0KtPG6mR2yQ      4   
207160   secsGLdQOaJPATB9SnHlew 2014-09-18  enDTy9GkIeMrgih4wLCSIw      5   
207823   JyfJYQURfRuDvrWRlDDELw 2011-12-23  AxVL5PEhP1X9N6s7eROhgA      1   
232364   259TmptwDUjXWgPlKf52Tw 2014-08-11  79sUuON3QMBlcSiRrZq4Ag      4   
232441   259TmptwDUjXWgPlKf52Tw 2016-06-28  MUy6TAlOC5iNCssjVWFK6A      5   
240620   q0aoi6vsi9TBU5-k3FfOOw 2016-04-05  _vRGe1nVJPPzA-RC86FQgw      5   
269381   7Mgly28cnzY-2py2ClEjiA 2015-03-02  wWcLsxgwkPgfjAa3NcXH3Q      5   
285910   QTxEdBIM9VXcyBnK3ucKUg 2014-06-15  ioiFazM87HnAqmOE-dnLlQ      1   
288891   VdokKATVykyF3YxtzN2raA 2013-04-18  -DnDYIOCbqUP1DmmhFcx4w      5   
320553   YCKOz3ja7i3nbavwdf6lWQ 2014-08-26  q1-9nC0OWaYBFBmZYLqUWw      4   
321727   gMq4GXSl21p2Z3PIQm3aqQ 2012-09-13  WkmO08ywlLiNUPjwvWPmfg      4   
341875   ff_QnODW0vqNWZPSTVQTsQ 2014-01-29  YF9-2WLn6_FUaSPkAVVFUA      5   
342349   HAMKYsgtJ6fEBUZzTMhZwA 2016-03-26  KNeuu0MBrlVEzrZx_Afv7A      3   
362342   Ijs6Fjhs-RETgbWRQTXHTw 2011-12-11  k6Jgo-O3VbOHO70_Qu5ryg      2   
381788   dcd3C1gWv-vVdQ9XYV8Ubw 2010-09-20  nPQ7zvWPhqT0a5-TKNmxwg      3   
...                         ...        ...                     ...    ...   
2223418  ghnQrKVEjiWZ3UE1XDzJUQ 2015-07-29  QwI1Pis5_HIcPyRPdlwcWw      4   
2223455  ghnQrKVEjiWZ3UE1XDzJUQ 2015-12-28  B4SaZixLmUWQUlI_rsEB6Q      4   
2223501  ghnQrKVEjiWZ3UE1XDzJUQ 2016-05-02  0RvDwXWAdvftuRPRCNhgYQ      1   
2223511  ghnQrKVEjiWZ3UE1XDzJUQ 2016-06-07  XhVtOaJpDF24HUOpijwoIQ      5   
2226999  MN0OLvP8i0SirMJ7AMuMzA 2015-05-03  _sR3sYEn2AXsKtVOkmj29Q      5   
2240369  phJiywlkjztj9jQETBO-Uw 2013-08-17  xjGLGS8IlxgQLdMdJxC3Hw      1   
2253827  -ZTR_3FFPdFCV9l_Fq_OUA 2015-03-25  yrIpBkKDiT7dzHT_Onx7eA      4   
2253945  _DZfwU9dE9Q5I1-1jiM1vQ 2016-01-28  xKXMxh2hHk08IH2MY2ZGQA      4   
2261352  NKyOOV0iOQt1l892IzJfpg 2014-11-20  s27zFNvLWXwSISjDGe1IaA      4   
2265145  8NsscdE7eiLCTq3sW2OqFw 2015-11-18  nAt_DuvtdJxvURrIou1MAA      5   
2273516  oPpA0LHMWJht4lMFgfYwhA 2014-05-01  VCLaMyDPbcZQ6TIF5ZxuUw      3   
2299048  aastrx4llH7kBL97DWfsXA 2014-05-24  mVTg7MKFiMY2fF5uKNl15A      4   
2299049  aastrx4llH7kBL97DWfsXA 2014-05-27  RQ3FNKfIA2j7Op-Aq5H8Aw      3   
2304654  d_QoB6QfsGVqSzB3cGO_tA 2016-06-03  EE5gXT0rKF7oWLy865-a4A      5   
2307956  zyvujnBGGpG3WWiDPkil9w 2015-03-15  atOjw1zcxiokrZk-De2d2A      4   
2336838  shA3pC1JuAkXl28_LVcp7w 2014-01-28  5kscHO1DxswrrsN7HBkIbQ      5   
2346945  JOX4ZgSQ5f9eRtSerJlzSg 2015-12-04  lNEDhO_ShQ9l3Y2YSixZNQ      5   
2476644  WmNL8V3mo4xtOXcIxqTsVA 2015-03-23  dIjy9XukgaxvZwrE0x0ibw      4   
2497589  m35p1HyBUl5tjxnN3atVTA 2014-12-05  BgRrbuWjuvLH2AeqAl99ig      2   
2510233  lGqTeEGIEp9aj5Tz0kyu_w 2015-02-26  c6M3TdLAMry1xUsykvNAMg      5   
2592528  Ghol-DN

In [154]:
pd.DataFrame(review.groupby(['user_id'], sort=True).size().rename('counts'))

counts
user_id                       
---teJGnwK07UO6_oJfbRw       1
--0HEXd4W6bJI8k7E0RxTA       2
--0KsjlAThNWua2Pr4HStQ       6
--0mI_q_0D1CdU4P_hoImQ       1
--106arHH4D3fLenTl3YZA       1
--1Y03CEKR3WDbBjYnsW7A       1
--2HuVGKHBpUBoOE4C7Qtg       4
--2QZsyXGz1OhiD4-0FQLQ       3
--3-eGXlu61VQZZy-JviNw       5
--37T3V10ZeoJ1Il7D2Wzw       1
--3lusamlL5SbQbnqZ4IUA       1
--4TkB_iDShmg41Y_QW9nw      35
--4fX3LBeXoE88gDTK6TKQ      16
--519Rh5sTtkoUraGzAaKQ       3
--5SlEtHW4LEbS0mEUMwag       1
--65q1FpAL_UQtVZ2PTGew     151
--7-XhUKUK3BCUqHHc2ENA       1
--7266Nwi6RXjKNNKeoFMQ       3
--80yFOfe6nZKLhxTMZjEg       1
--82_AVgRBsLw6Dhy8sEnA       1
--8A9o_NeGyt_3kzlXtSdg       3
--8CfkbFfWMXsPPq2cwqQQ       1
--8WbseBk1NjfPiZWjQ-XQ       4
--9HuvEtLhp21SeIEEltnA       3
--9jRaeY1xK-2l9r9fVQWA      14
--Abx_EDHkKd8Y7t8U5kHQ       1
--AfacevQEmlfHd8Ed1NWg      33
--AqBuo717RyacADExYbSA       8
--BGW_TY55SH-9OiHmtitg       1
--BnK7rx64Gu9KAb2OjMxQ      18
...                        ...
zzn1dAYKc8KEZk6kBisl6A       1
zznPTkBKMznV8RdqvDX5Jw       2
zznUhVBuxM-Hl5_2P6bO_Q       1
zznlyyjRcsxLHj5pDMLL2A       1
zzntUWPa2pGOpPBe7VZ5AQ       1
zzoe0-0JkltGZbQMdZ8x4w       1
zzotrBOhamZZ1Q6Mn0EbOQ       1
zzpT-Mfyh4jEjdqDerSVnQ       3
zzpgN3i0I7q239rcpteI2Q       1
zzqQ1PpMwURVaQsr2m3-yg       9
zzqlE_xZK-ls6wQntT6fKQ       1
zzr0gcwOWFGBBqQ11_xPOQ       6
zzrFOtxRt5oVRs_frrQ2iQ       1
zzrNDzW1GxgaWTBD2yeaaQ       1
zztT9KKp0rmJRMeMl4CqzQ       4
zzu0dZXo10IxEJduMVHgTg       2
zzu_HGDO-nv0gHivZrKdpQ       6
zzunUJtZqhcxDB3plnSFxg       1
zzv762b871Wp3SVxsFQqHw       1
zzvGAKv1D1fdUwCvXEeQnA       3
zzvflqB4xAcKFbIbS6HSOw      12
zzvrDmRAYdvIVRGszSFPGQ       2
zzwu0nG0MCUP43zNSr5k-A       1
zzx-JA0qEz5eTO3GMv99NA       1
zzxXZcJ89sSyJEzoNo6slw       1
zzx_41wDUNxpcgrdOtERvw      13
zzyeArRv6I5HpEJlOCOPAQ       3
zzyoUJV5QTUEuuVoICcdYQ       2
zzytqT9s0WS-d4WgLYmLoA       1
zzzfn8ETe8A4kp4HnTK86g       2

[686556 rows x 1 columns]

In [152]:
pd.DataFrame(review.groupby(['user_id']).size().rename('counts').reset_index())

user_id  counts
0       ---teJGnwK07UO6_oJfbRw       1
1       --0HEXd4W6bJI8k7E0RxTA       2
2       --0KsjlAThNWua2Pr4HStQ       6
3       --0mI_q_0D1CdU4P_hoImQ       1
4       --106arHH4D3fLenTl3YZA       1
5       --1Y03CEKR3WDbBjYnsW7A       1
6       --2HuVGKHBpUBoOE4C7Qtg       4
7       --2QZsyXGz1OhiD4-0FQLQ       3
8       --3-eGXlu61VQZZy-JviNw       5
9       --37T3V10ZeoJ1Il7D2Wzw       1
10      --3lusamlL5SbQbnqZ4IUA       1
11      --4TkB_iDShmg41Y_QW9nw      35
12      --4fX3LBeXoE88gDTK6TKQ      16
13      --519Rh5sTtkoUraGzAaKQ       3
14      --5SlEtHW4LEbS0mEUMwag       1
15      --65q1FpAL_UQtVZ2PTGew     151
16      --7-XhUKUK3BCUqHHc2ENA       1
17      --7266Nwi6RXjKNNKeoFMQ       3
18      --80yFOfe6nZKLhxTMZjEg       1
19      --82_AVgRBsLw6Dhy8sEnA       1
20      --8A9o_NeGyt_3kzlXtSdg       3
21      --8CfkbFfWMXsPPq2cwqQQ       1
22      --8WbseBk1NjfPiZWjQ-XQ       4
23      --9HuvEtLhp21SeIEEltnA       3
24      --9jRaeY1xK-2l9r9fVQWA      14
25      --Abx_EDHkKd8Y7t8U5kHQ       1
26      --AfacevQEmlfHd8Ed1NWg      33
27      --AqBuo717RyacADExYbSA       8
28      --BGW_TY55SH-9OiHmtitg       1
29      --BnK7rx64Gu9KAb2OjMxQ      18
...                        ...     ...
686526  zzn1dAYKc8KEZk6kBisl6A       1
686527  zznPTkBKMznV8RdqvDX5Jw       2
686528  zznUhVBuxM-Hl5_2P6bO_Q       1
686529  zznlyyjRcsxLHj5pDMLL2A       1
686530  zzntUWPa2pGOpPBe7VZ5AQ       1
686531  zzoe0-0JkltGZbQMdZ8x4w       1
686532  zzotrBOhamZZ1Q6Mn0EbOQ       1
686533  zzpT-Mfyh4jEjdqDerSVnQ       3
686534  zzpgN3i0I7q239rcpteI2Q       1
686535  zzqQ1PpMwURVaQsr2m3-yg       9
686536  zzqlE_xZK-ls6wQntT6fKQ       1
686537  zzr0gcwOWFGBBqQ11_xPOQ       6
686538  zzrFOtxRt5oVRs_frrQ2iQ       1
686539  zzrNDzW1GxgaWTBD2yeaaQ       1
686540  zztT9KKp0rmJRMeMl4CqzQ       4
686541  zzu0dZXo10IxEJduMVHgTg       2
686542  zzu_HGDO-nv0gHivZrKdpQ       6
686543  zzunUJtZqhcxDB3plnSFxg       1
686544  zzv762b871Wp3SVxsFQqHw       1
686545  zzvGAKv1D1fdUwCvXEeQnA       3
686546  zzvflqB4xAcKFbIbS6HSOw      12
686547  zzvrDmRAYdvIVRGszSFPGQ       2
686548  zzwu0nG0MCUP43zNSr5k-A       1
686549  zzx-JA0qEz5eTO3GMv99NA       1
686550  zzxXZcJ89sSyJEzoNo6slw       1
686551  zzx_41wDUNxpcgrdOtERvw      13
686552  zzyeArRv6I5HpEJlOCOPAQ       3
686553  zzyoUJV5QTUEuuVoICcdYQ       2
686554  zzytqT9s0WS-d4WgLYmLoA       1
686555  zzzfn8ETe8A4kp4HnTK86g       2

[686556 rows x 2 columns]

In [161]:
user.iloc[[0]]

average_stars                                        compliments  \
0           4.14  {'writer': 9, 'more': 3, 'cute': 15, 'cool': 7...   

          elite  fans                                            friends  \
0  [2005, 2006]    69  [rpOyqD_893cqmDAtJLbdog, 4U9kSBLuBDU391x6bxU-Y...   

     name  review_count  type                 user_id  \
0  Russel           108  user  18kPq7GPye-YQ3LyKyAZPw   

                                        votes yelping_since  
0  {'funny': 167, 'cool': 246, 'useful': 282}       2004-10

In [162]:
checkin.head(1)

business_id                                       checkin_info  \
0  cE27W9VPgO88Qxe4ol6y_g  {'17-3': 1, '18-4': 1, '9-5': 1, '13-0': 1, '1...   

      type  
0  checkin

In [23]:
user.compliments.iloc[0]

{'cool': 76,
 'cute': 15,
 'funny': 11,
 'hot': 48,
 'more': 3,
 'note': 20,
 'photos': 14,
 'plain': 25,
 'profile': 8,
 'writer': 9}

In [77]:
len(business.business_id.unique())

85901

In [294]:
review.columns

Index(['business_id', 'date', 'review_id', 'stars', 'text', 'type', 'user_id',
       'votes'],
      dtype='object')

In [299]:
len(set(review.business_id))

85539

In [10]:
len(review.business_id.unique())

85539

In [114]:
review.head()

business_id       date               review_id  stars  \
0  5UmKMjUEUNdYWqANhGckJw 2012-08-01  Ya85v4eqdd6k9Od8HbQjyA      4   
1  5UmKMjUEUNdYWqANhGckJw 2014-02-13  KPvLNJ21_4wbYNctrOwWdQ      5   
2  5UmKMjUEUNdYWqANhGckJw 2015-10-31  fFSoGV46Yxuwbr3fHNuZig      5   
3  5UmKMjUEUNdYWqANhGckJw 2015-12-26  pVMIt0a_QsKtuDfWVfSk2A      3   
4  5UmKMjUEUNdYWqANhGckJw 2016-04-08  AEyiQ_Y44isJmNbMTyoMKQ      2   

                                                text    type  \
0  Mr Hoagie is an institution. Walking in, it do...  review   
1  Excellent food. Superb customer service. I mis...  review   
2  Yes this place is a little out dated and not o...  review   
3  PROS: Italian hoagie was delicious.  Friendly ...  review   
4  First the only reason this place could possibl...  review   

                  user_id                                 votes  
0  PUFPaY9KxDAcGqfsorJp3Q  {'cool': 0, 'funny': 0, 'useful': 0}  
1  Iu6AxdBYGR4A0wspR9BYHA  {'cool': 0, 'funny': 0, 'useful': 0}  
2  auESFwWvW42h6alXgFxAXQ  {'cool': 0, 'funny': 1, 'useful': 1}  
3  qiczib2fO_1VBG8IoCGvVg  {'cool': 0, 'funny': 0, 'useful': 0}  
4  qEE5EvV-f-s7yHC0Z4ydJQ  {'cool': 0, 'funny': 0, 'useful': 1}

In [10]:
review.stars.unique()

array([4, 5, 3, 2, 1], dtype=int64)

In [108]:
tip[tip.likes>10].values

array([['iTLwGhF-NyHGrgoYzGK7xw', Timestamp('2015-09-17 00:00:00'), 11,
        'License photographer assured me her photos always came out great. That is false! My head looks like something Super Mario stomps on. Greeeeeaaaaaat.',
        'tip', 'WmAyExqSWoiYZ5XEqpk_Uw']], dtype=object)

In [106]:
business[business.business_id==tip[tip.likes>10].values[0][0]]

attributes             business_id  \
20925  {'Accepts Credit Cards': True}  iTLwGhF-NyHGrgoYzGK7xw   

                                              categories       city  \
20925  [Departments of Motor Vehicles, Public Service...  Las Vegas   

                                            full_address  \
20925  8250 W Flamingo Rd\nSpring Valley\nLas Vegas, ...   

                                                   hours   latitude  \
20925  {'Tuesday': {'open': '08:00', 'close': '17:00'...  36.115587   

        longitude                          name    neighborhoods  open  \
20925 -115.270839  Department of Motor Vehicles  [Spring Valley]  True   

       review_count  stars state      type  
20925           189    3.0    NV  business

dfreview.apply(lambda row: 'none' in row.values, axis=1).any()

output is False

In [33]:
business[business.longitude>=0].city.shape

(1074,)

In [37]:
business_checkin = business.merge(checkin, left_on='business_id', right_on='business_id')

In [39]:
business_checkin.shape

(61049, 17)

In [40]:
print(business.shape, checkin.shape)

(85901, 15) (61049, 3)


In [43]:
business_checkin.head()

attributes             business_id  \
0                            {'Good for Kids': True}  cE27W9VPgO88Qxe4ol6y_g   
1  {'Music': {'dj': False}, 'Take-out': True, 'Co...  mVHrayjG3uZ_RLHkLj-AMg   
2  {'Accepts Credit Cards': True, 'Good for Kids'...  mYSpR_SLPgUVymYOvTQd_Q   
3  {'Music': {'dj': False}, 'Take-out': True, 'Co...  KayYbHCt-RkbGcPdGOThNg   
4                     {'Accepts Credit Cards': True}  b12U9TFESStdy7CsTtcOeg   

                                          categories         city  \
0                     [Active Life, Mini Golf, Golf]  Bethel Park   
1  [Bars, American (New), Nightlife, Lounges, Res...     Braddock   
2                                [Active Life, Golf]     Braddock   
3  [Bars, American (Traditional), Nightlife, Rest...     Carnegie   
4                   [Auto Repair, Automotive, Tires]     Carnegie   

                                      full_address  \
0          1530 Hamilton Rd\nBethel Park, PA 15234   
1              414 Hawkins Ave\nBraddock, PA 15104   
2            1000 Clubhouse Dr\nBraddock, PA 15104   
3  141 Hawthorne St\nGreentree\nCarnegie, PA 15106   
4          718A Hope Hollow Rd\nCarnegie, PA 15106   

                                               hours   latitude  longitude  \
0                                                 {}  40.354115 -80.014660   
1  {'Friday': {'close': '20:00', 'open': '10:00'}...  40.408830 -79.866211   
2  {'Thursday': {'close': '20:00', 'open': '11:00...  40.403405 -79.855782   
3  {'Monday': {'close': '02:00', 'open': '11:00'}...  40.415486 -80.067549   
4  {'Monday': {'close': '18:00', 'open': '07:30'}...  40.394588 -80.084454   

                          name neighborhoods   open  review_count  stars  \
0     Cool Springs Golf Center            []  False             5    2.5   
1                Emil's Lounge            []   True            26    4.5   
2         Grand View Golf Club            []   True             3    5.0   
3        Alexion's Bar & Grill   [Greentree]   True            23    4.0   
4  Flynn's Tire & Auto Service            []   True             9    2.5   

  state    type_x                                       checkin_info   type_y  
0    PA  business  {'14-6': 1, '17-6': 1, '17-3': 1, '18-4': 1, '...  checkin  
1    PA  business  {'6-2': 1, '13-1': 1, '9-4': 3, '15-0': 1, '12...  checkin  
2    PA  business  {'13-5': 1, '17-3': 1, '5-4': 1, '11-4': 1, '5...  checkin  
3    PA  business  {'19-5': 1, '17-4': 1, '17-3': 2, '13-2': 2, '...  checkin  
4    PA  business  {'5-5': 2, '8-2': 1, '8-6': 1, '5-1': 2, '13-2...  checkin

In [18]:
business[business.state =='']

attributes  \
81078  {'Good for Kids': True, 'Take-out': True, 'Dri...   

                  business_id                categories      city  \
81078  HvP4Tvx9mihceKb2gnbhgw  [Fast Food, Restaurants]  Montreal   

                                            full_address hours   latitude  \
81078  Pierre Elliott Trudeau Airport\nSaint-Laurent\...    {}  45.456136   

       longitude         name            neighborhoods  open  review_count  \
81078 -73.752866  Burger King  [Saint-Laurent, Dorval]  True             4   

       stars state      type  
81078    2.0        business

In [16]:
business.categories[3]

['Bars', 'American (New)', 'Nightlife', 'Lounges', 'Restaurants']

In [12]:
business.attributes.iloc[0]

{'Accepts Credit Cards': True,
 'Alcohol': 'none',
 'Ambience': {'casual': False,
  'classy': False,
  'divey': False,
  'hipster': False,
  'intimate': False,
  'romantic': False,
  'touristy': False,
  'trendy': False,
  'upscale': False},
 'Attire': 'casual',
 'Caters': False,
 'Delivery': False,
 'Drive-Thru': False,
 'Good For': {'breakfast': False,
  'brunch': False,
  'dessert': False,
  'dinner': False,
  'latenight': False,
  'lunch': False},
 'Good For Groups': True,
 'Good for Kids': True,
 'Has TV': False,
 'Noise Level': 'average',
 'Outdoor Seating': False,
 'Parking': {'garage': False,
  'lot': False,
  'street': False,
  'valet': False,
  'validated': False},
 'Price Range': 1,
 'Take-out': True,
 'Takes Reservations': False,
 'Waiter Service': False}

In [139]:
len(review.text)

2685066

In [79]:
2685066-2679293

5773

In [80]:
dfreview.text[dfreview.duplicated(['text', 'business_id'], keep=False)].count()

6684

In [109]:
review.duplicated([review.columns.difference(['votes'])][0].values,keep=False).sum()

0

In [75]:
review.duplicated(['business_id','user_id'],keep=False).sum()

157088

In [71]:
review.duplicated(['business_id', 'text'],keep=False).sum()

6684

In [11]:
review.ix[2,'date']

Timestamp('2015-10-31 00:00:00')

In [110]:
[review['votes'][i]['cool'] for i in range(19,30)]

[1, 0, 0, 1, 0, 0, 2, 0, 4, 0, 0]

In [15]:
#df.groupby('business_id').count().sort_values('date', ascending = False)

In [46]:
review_tip = review.merge(tip, left_on=['business_id', 'user_id'], right_on=['business_id', 'user_id'])

In [57]:
review_tip[review_tip.likes>1].iloc[1].votes

{'cool': 2, 'funny': 6, 'useful': 5}

In [61]:
photosmapfilepath = 'yelp_photos/photo_id_to_business_id.json'

In [13]:
# with open('yelp_photos/photo_id_to_business_id.json') as file:
#         photos = pd.read_json(file)

In [65]:
checkin.shape

(61049, 3)

In [67]:
checkin.head(1)

business_id                                       checkin_info  \
0  cE27W9VPgO88Qxe4ol6y_g  {'14-6': 1, '17-6': 1, '17-3': 1, '18-4': 1, '...   

      type  
0  checkin

In [21]:
(checkin.loc[1,'checkin_info'])

{'10-4': 3,
 '10-5': 1,
 '11-5': 1,
 '12-1': 1,
 '12-5': 2,
 '13-1': 1,
 '14-4': 2,
 '15-0': 1,
 '15-1': 1,
 '15-4': 2,
 '16-2': 1,
 '16-3': 1,
 '17-4': 1,
 '21-4': 1,
 '22-5': 1,
 '6-2': 1,
 '8-2': 1,
 '8-4': 1,
 '9-4': 3,
 '9-5': 2}

In [84]:
checkin.checkin_info.apply(lambda row: len(row))

0          9
1         20
2          8
3         44
4          5
5          6
6          5
7          3
8         14
9         21
10        12
11        44
12        13
13        12
14         7
15         6
16        31
17        48
18         6
19         3
20        23
21        93
22        47
23        20
24        68
25        24
26        64
27        47
28        75
29         9
        ... 
61019     13
61020     20
61021      3
61022      4
61023      5
61024      6
61025      6
61026      4
61027    122
61028      5
61029      8
61030     43
61031     15
61032     21
61033      3
61034      3
61035      4
61036      6
61037     10
61038      4
61039      9
61040     28
61041    104
61042     71
61043      4
61044      9
61045      9
61046     86
61047     23
61048     12
Name: checkin_info, dtype: int64

In [83]:
checkin.checkin_info.apply(lambda row: len(row)).min()

3

In [80]:
len({1:2, 3:5})

2

In [87]:
business.attributes[1000]

{'Accepts Credit Cards': True,
 'Alcohol': 'full_bar',
 'Ambience': {'casual': False,
  'classy': False,
  'divey': False,
  'hipster': False,
  'intimate': False,
  'romantic': False,
  'touristy': False,
  'trendy': False,
  'upscale': False},
 'Attire': 'casual',
 'Delivery': False,
 'Good For': {'breakfast': False,
  'brunch': False,
  'dessert': False,
  'dinner': False,
  'latenight': False,
  'lunch': True},
 'Good For Dancing': False,
 'Good For Groups': True,
 'Good for Kids': False,
 'Happy Hour': True,
 'Has TV': True,
 'Noise Level': 'loud',
 'Outdoor Seating': False,
 'Parking': {'garage': True,
  'lot': False,
  'street': False,
  'valet': False,
  'validated': False},
 'Price Range': 1,
 'Take-out': True,
 'Takes Reservations': False,
 'Waiter Service': True}

In [88]:
business.columns

Index(['attributes', 'business_id', 'categories', 'city', 'full_address',
       'hours', 'latitude', 'longitude', 'name', 'neighborhoods', 'open',
       'review_count', 'stars', 'state', 'type'],
      dtype='object')

In [428]:
a = business.groupby(business.city)

In [433]:
pd.DataFrame({'count':a.size(), 'avg_stars': a.stars.mean()}).reset_index()

city  avg_stars  count
0                                  4.500000      2
1                  110 Las Vegas   5.000000      1
2                     Ahwahtukee   5.000000      1
3                      Ahwatukee   3.750000     12
4    Ahwatukee Foothills Village   5.000000      1
5                      Allegheny   5.000000      1
6                      Allentown   3.500000      2
7                          Anjou   3.318182     11
8                         Anthem   3.696296    135
9                Apache Junction   3.596774    186
10                     Arlington   3.750000      2
11             Arlington Heights   4.500000      1
12                     Asheville   4.500000      1
13                     Aspinwall   3.916667     12
14                        Avalon   3.500000      4
15                      Avondale   3.532787    427
16                   Baie-D'urfe   3.500000      1
17                       Balerno   2.000000      1
18                    Banksville   3.000000      1
19                  Beaconsfield   4.200000      5
20                      Bellevue   3.833333     24
21                       Bellvue   3.833333      3
22                       Belmont   3.759036     83
23                   Bethel Park   2.666667      3
24                    Bietigheim   3.625000      4
25             Black Canyon City   3.333333      3
26                    Blainville   3.750000      2
27                       Blawnox   4.250000      2
28                    Bloomfield   4.000000      1
29                  Blue Diamond   4.500000      1
..                           ...        ...    ...
414                  W Summerlin   3.000000      1
415                      Waddell   3.375000      8
416                    Waldbronn   4.166667      9
417               Water of Leith   4.500000      1
418                     Waterloo   3.496633    297
419                     Waunakee   3.597561     41
420                       Waxhaw   4.000000      5
421                   Weddington   4.500000      4
422                   Weingarten   4.090909     11
423           Weingarten (Baden)   4.166667      3
424                Wesley Chapel   3.625000      8
425               West Homestead   3.458333     12
426               West Las Vegas   2.500000      1
427                 West Mifflin   3.666667     36
428                    Westmount   3.745763     59
429         Westworld Scottsdale   3.000000      1
430                    Whitehall   3.000000      1
431                      Whitney   4.500000      1
432                   Wickenburg   3.794872     78
433                  Wilkinsburg   3.444444      9
434                      Windsor   3.400000     10
435                     Wittmann   3.500000      1
436                        Wörth   4.000000      1
437               Wörth am Rhein   4.062500      8
438                    Youngtown   4.000000     30
439                     glendale   4.333333      3
440                      phoenix   3.875000      4
441               pleasant hills   2.000000      1
442                   scottsdale   4.666667      3
443               Île des Soeurs   4.000000      1

[444 rows x 3 columns]

In [435]:
business.iloc[0]

attributes       {'Take-out': True, 'Has TV': False, 'Noise Lev...
business_id                                 5UmKMjUEUNdYWqANhGckJw
categories                                [Fast Food, Restaurants]
city                                                    Dravosburg
full_address          4734 Lebanon Church Rd\nDravosburg, PA 15034
hours            {'Tuesday': {'open': '11:00', 'close': '21:00'...
latitude                                                   40.3543
longitude                                                 -79.9007
name                                                     Mr Hoagie
neighborhoods                                                   []
open                                                          True
review_count                                                     7
stars                                                          3.5
state                                                           PA
type                                                      busi

# Find list of users with all available reviews
## if there is enough data, some ideas can be implemented

In [166]:
user[['user_id', 'review_count']]

user_id  review_count
0       18kPq7GPye-YQ3LyKyAZPw           108
1       rpOyqD_893cqmDAtJLbdog          1292
2       4U9kSBLuBDU391x6bxU-YA           395
3       fHtTaujcyKvXglE33Z5yIw            11
4       SIBCL7HBkrP4llolm4SC2A            66
5       8J4IIYcqBlFch8T90N923A          1632
6       ysYmC-ufbdmVEX9yAv-VEQ            55
7       WPOKvkacSKHx_bIG1alFiA            60
8       UTS9XcT14H2ZscRIf0MYHQ           102
9       1blidZhgxDVSBuJ_dbTU6g             1
10      qL7Astun3i7qwr2IL5iowA            20
11      ZWOj6LmzwGvMDh-A85EOtA           984
12      HDQixQ-WZEV0LVPJlIGQeQ            48
13      9i8tV1cWR3nSiOTtg4OEAg           122
14      bWNn5RzthU9-yo5VYQB20g            24
15      xu_KA5FxXfb5-vx0uz_BNg             5
16      5OlCB4cJ3CUksr3ONw6Ezw             3
17      e-hBnxMWmzSut-BlyxR2KQ            20
18      E_DJ1nTDPt9Qk2abWFTefQ            13
19      LCbwNuo4r-WdAmw6LeGPog             1
20      xOQVHYN1roRZKpLvAT-a2A             2
21      Z28E8hEXDoLZUuPcZvnzcg            14
22      0ZtOuzEA3k8XjSaf6MgR-w            55
23      J817Cq7qJ9bYvJVFHlUgCw            22
24      Yt4bcTwDMDCGSD2eQHxazw             3
25      fBEydVU1uTsfGvFI2CaEaw             6
26      o625WyBtvJ_G3s0FRr6RmQ            10
27      Il4hru5zwULRgCxqOcSfyg            25
28      ryu1gXViHZJ38-o9fr9vVw            18
29      LF73iBr-ynbAi-swAQbGCQ             7
...                        ...           ...
686526  VITAFPhy5rkYkNexQiSjTw             9
686527  QkFsYvuICPVpFba7DPA_tg             7
686528  TPQM-PeqmbOjWZTWOntUmQ             1
686529  YROaWkoihQ6xflZ26twKXA             4
686530  zcgNlDNoe-O8t4cmO4_zxQ             6
686531  QHN8WM7C6qci35G0Tr7C0g             2
686532  UjaM7Sm8_Xq2fWB04SNdSA             3
686533  A03eMT1ukzlh-_kZ34ExCw             2
686534  0GCws7DKxvdtvF4c_qcEUw             2
686535  HNv_DXLbvUGh976mHXzbsA             1
686536  6Dwqk9LU6bv9ED5K-mj1zQ             2
686537  o5ZwzvxQLJ8U4dRERm1OdQ             3
686538  B5FbMpgMnUonweq59JdoZA             5
686539  EPNrLMgIrdlzIaXojWb2_w             4
686540  OhR-zDBVoGC7GjZ8h3MZjw             2
686541  2COS25ubc03qZUJlb9co7w            10
686542  d4hcITpuuwMbt1lk5bfioQ             6
686543  HbALYkupeF3P0HH6OHa-VQ            13
686544  uAcPATQyIwb--qZ8gb3Npg             1
686545  lCHEFhRPKupyy5p0GC7mhA             1
686546  l5GTCveGC2SP1adAeDb5Sw             3
686547  cSVme2vhuSV9t-5lGQnN8w             2
686548  AjhWrw9ltv8Ehe5CbEMjWg             2
686549  MMGM5YC49IBA-HWwjUNBfQ            13
686550  S47mbC8NcD5LA6rC_03RoQ             1
686551  edetLs2KhHLCiJilqjdeUQ             4
686552  q0tofG0bWFv8ou0ea4aAiA            51
686553  JV-CHnvX6qrQzRQOL2yxCA             3
686554  uRM8uxxbhum_Bedx1c1frw             1
686555  DL0S4Ro4KY77akGPbEkrug             1

[686556 rows x 2 columns]

In [178]:
review_by_user = pd.DataFrame(review.groupby(['user_id']).size().rename('reviews').reset_index())

In [188]:
testing_review_counts = review_by_user.merge(user[['user_id', 'review_count', 'elite']], left_on='user_id', right_on='user_id')
testing_review_counts

user_id  reviews  review_count  \
0       ---teJGnwK07UO6_oJfbRw        1             2   
1       --0HEXd4W6bJI8k7E0RxTA        2            34   
2       --0KsjlAThNWua2Pr4HStQ        6            60   
3       --0mI_q_0D1CdU4P_hoImQ        1            31   
4       --106arHH4D3fLenTl3YZA        1             1   
5       --1Y03CEKR3WDbBjYnsW7A        1             2   
6       --2HuVGKHBpUBoOE4C7Qtg        4            76   
7       --2QZsyXGz1OhiD4-0FQLQ        3             7   
8       --3-eGXlu61VQZZy-JviNw        5             6   
9       --37T3V10ZeoJ1Il7D2Wzw        1            14   
10      --3lusamlL5SbQbnqZ4IUA        1             1   
11      --4TkB_iDShmg41Y_QW9nw       35            44   
12      --4fX3LBeXoE88gDTK6TKQ       16            26   
13      --519Rh5sTtkoUraGzAaKQ        3             8   
14      --5SlEtHW4LEbS0mEUMwag        1             1   
15      --65q1FpAL_UQtVZ2PTGew      151           205   
16      --7-XhUKUK3BCUqHHc2ENA        1             5   
17      --7266Nwi6RXjKNNKeoFMQ        3            64   
18      --80yFOfe6nZKLhxTMZjEg        1             8   
19      --82_AVgRBsLw6Dhy8sEnA        1            18   
20      --8A9o_NeGyt_3kzlXtSdg        3             4   
21      --8CfkbFfWMXsPPq2cwqQQ        1            11   
22      --8WbseBk1NjfPiZWjQ-XQ        4            65   
23      --9HuvEtLhp21SeIEEltnA        3             5   
24      --9jRaeY1xK-2l9r9fVQWA       14            60   
25      --Abx_EDHkKd8Y7t8U5kHQ        1            47   
26      --AfacevQEmlfHd8Ed1NWg       33           124   
27      --AqBuo717RyacADExYbSA        8             8   
28      --BGW_TY55SH-9OiHmtitg        1             1   
29      --BnK7rx64Gu9KAb2OjMxQ       18            19   
...                        ...      ...           ...   
686526  zzn1dAYKc8KEZk6kBisl6A        1             9   
686527  zznPTkBKMznV8RdqvDX5Jw        2             2   
686528  zznUhVBuxM-Hl5_2P6bO_Q        1             5   
686529  zznlyyjRcsxLHj5pDMLL2A        1             4   
686530  zzntUWPa2pGOpPBe7VZ5AQ        1             2   
686531  zzoe0-0JkltGZbQMdZ8x4w        1             2   
686532  zzotrBOhamZZ1Q6Mn0EbOQ        1             1   
686533  zzpT-Mfyh4jEjdqDerSVnQ        3             3   
686534  zzpgN3i0I7q239rcpteI2Q        1             1   
686535  zzqQ1PpMwURVaQsr2m3-yg        9            31   
686536  zzqlE_xZK-ls6wQntT6fKQ        1             1   
686537  zzr0gcwOWFGBBqQ11_xPOQ        6             8   
686538  zzrFOtxRt5oVRs_frrQ2iQ        1             7   
686539  zzrNDzW1GxgaWTBD2yeaaQ        1            13   
686540  zztT9KKp0rmJRMeMl4CqzQ        4            40   
686541  zzu0dZXo10IxEJduMVHgTg        2           624   
686542  zzu_HGDO-nv0gHivZrKdpQ        6           116   
686543  zzunUJtZqhcxDB3plnSFxg        1             1   
686544  zzv762b871Wp3SVxsFQqHw        1             1   
686545  zzvGAKv1D1fdUwCvXEeQnA        3             4   
686546  zzvflqB4xAcKFbIbS6HSOw       12            36   
686547  zzvrDmRAYdvIVRGszSFPGQ        2            18   
686548  zzwu0nG0MCUP43zNSr5k-A        1            10   
686549  zzx-JA0qEz5eTO3GMv99NA        1             1   
686550  zzxXZcJ89sSyJEzoNo6slw        1             1   
686551  zzx_41wDUNxpcgrdOtERvw       13            14   
686552  zzyeArRv6I5HpEJlOCOPAQ        3             3   
686553  zzyoUJV5QTUEuuVoICcdYQ        2             2   
686554  zzytqT9s0WS-d4WgLYmLoA        1             3   
686555  zzzfn8ETe8A4kp4HnTK86g        2             4   

                                       elite  
0                                         []  
1                                         []  
2                               [2014, 2015]  
3                                         []  
4                                         []  
5                                         []  
6                                         []  
7                                         []  
8                                         []  
9                 

In [215]:
(testing_review_counts[(testing_review_counts.reviews == testing_review_counts.review_count) & (testing_review_counts.elite.map(len) >0)]).shape

(34, 4)

In [213]:
testing_review_counts[(testing_review_counts.reviews == testing_review_counts.review_count) & (testing_review_counts.elite.map(len) >0)]

user_id  reviews  review_count  \
419     -1Q1s_NMGjBLBULA8z_npg        1             1   
6260    -_3p-2J3onbOHKMltqTCdQ       27            27   
48282   3VIRjR2V29TOaI9u64hMig       62            62   
72150   5jODkdonMV0HbIIPik1RuA      131           131   
100740  8NlGnhOVjqQOSF4M4uiwdQ       29            29   
134649  BZSM6JL92epbwz-Oj3u8KQ       51            51   
155175  DTZ5z1pbXQG1_c-D3HGmlQ       28            28   
209137  IUWjTmXc3wLVaMHz33inaA       45            45   
214058  IwxFKXDokKZdiD5iUdJPgQ       99            99   
227475  KCQUgDM3vNuB5xn4ZXVo-A      280           280   
250614  MLNXzph9nlCXGiSuv5G5Yg      215           215   
252132  MUHoBFzwRoXpCpgYSN8KRA      156           156   
254175  MfQRz8He78Z8C4dYdGnFAg       38            38   
269704  O6yAy9-DOnaOv-eeF_E7Ow       36            36   
271737  OJ4FQEvd2h8QXCMHwCostg      100           100   
278536  OwCEsC0-oRxsSpTHZ0rrvA       38            38   
297573  QhtTrWSAt-KGyEnkSZZb4g       84            84   
297662  QiPECcsRMFjSkojdrvnsqA       57            57   
302223  R8BGrz2N8pdCnf0nf2iv6Q       24            24   
303494  RFpApVLkcVTU-aC6UteiLA       44            44   
363043  Wn4cr3qd67lYL_aya4I2TA       29            29   
407015  _tQtb7ZAJhiR9K8dWmf22w       24            24   
414974  ae16UwglIQ0FKdpPSpAH_g       73            73   
418701  b-4AaWfqfxndGlval5eONg       38            38   
434003  cSAACR699NuLMqBPuKL8nw      112           112   
466412  fTlX6YCEY8uwNCWzGtvx-A       50            50   
485580  hFv76gq-VKR7FLl_HZSn9Q       19            19   
500998  ihd9W-rqHVBo5bsz84adlA       21            21   
512923  jopndPrv-H5KW2CfScnw9A      219           219   
528623  lHPFFu5Hq2G1D17q1aU8Bg       45            45   
599060  rprUFN9Qnx5hC8Gs1-JIyQ       80            80   
601079  s10b9Ozb5RPKrou6T0v57A       66            66   
657106  xFFmQTr0srCsnXiupIXT2A       39            39   
684894  zptl58gf928g4JOOP2N0-w       55            55   

                                       elite  
419                       [2005, 2006, 2007]  
6260                                  [2012]  
48282                                 [2010]  
72150                           [2015, 2016]  
100740                                [2016]  
134649                                [2016]  
155175                          [2015, 2016]  
209137                          [2010, 2011]  
214058                    [2012, 2013, 2014]  
227475  [2011, 2012, 2013, 2014, 2015, 2016]  
250614              [2013, 2014, 2015, 2016]  
252132        [2008, 2009, 2010, 2011, 2012]  
254175                          [2014, 2015]  
269704                          [2015, 2016]  
271737                    [2011, 2012, 2013]  
278536                          [2011, 2012]  
297573                                [2016]  
297662                          [2015, 2016]  
302223                                [2016]  
303494                          [2014, 2015]  
363043                                [2012]  
407015                                [2014]  
414974                          [2010, 2011]  
418701                                [2016]  
434003                          [2012, 2013]  
466412                                [2016]  
485580                                [2015]  
500998                                [2015]  
512923        [2009, 2010, 2011, 2012, 2013]  
528623                          [2013, 2014]  
599060                                [2012]  
601079                                [2016]  
657106                          [2012, 2013]  
684894                                [2016]

In [204]:
testing_review_counts[(testing_review_counts.reviews == testing_review_counts.review_count) & (testing_review_counts.reviews > 19)]

user_id  reviews  review_count         elite
296     -0m1syFkZ8cQpRhWV64MNg       22            22            []
1416    -7JDe3dDexgYpcUOQ-tlaQ       21            21            []
3029    -H9XDexls_S9qBbzMUaNTw       29            29            []
4265    -OUKIatcSXuF-40IHcarwA       20            20            []
4380    -P9SkvexGLSavupPHMxZsA       27            27            []
4658    -QkQ-fvJKPCcnlQML6xokA       22            22            []
4839    -RfpEgUKX_IDSRXM2m_sqQ       36            36            []
5410    -V8vKRi8wBw-epeRPZeY_A       25            25            []
5486    -VhBUi5p4t1TnrwHuPxfhg       22            22            []
6260    -_3p-2J3onbOHKMltqTCdQ       27            27        [2012]
7352    -fhkLP6B7BDAZ1jBrAfwwg       62            62            []
7908    -iwV51Ygo3-xlc1Zqu7n9Q       26            26            []
8816    -oQ8t2USRuHM1-HuRdMQmw       34            34            []
10710   0-CJuByDISHBDa-IlOXb2g       40            40            []
10894   00PjBksrk0DYsG1LiJBnHQ       33            33            []
12153   07vARVikMIQkXMeIyzZnLg       20            20            []
13223   0EY9rt4QYSoHgTXlbPk8pw       29            29            []
17246   0bhUWCzzOkChsqMyDX-QQw       28            28            []
18719   0k5Gg9wEyi0GE6XOauueOw       27            27            []
19303   0ncNGqGiGAuNKUwZsrv1eg       20            20            []
23223   19qaAnRpolgII-xwOnrE0g       35            35            []
23723   1CpyQ1VkgmNwA6RxHoq2bA       40            40            []
24742   1IuIQAozG239xJqfBQfmxw       34            34            []
25456   1N-u1aavQN5hTUEPQOySyA       27            27            []
25696   1OTnkDxkVFjdOpePMiXv6A       32            32            []
27831   1aNyBVJe2hNSXrVY5zsR1Q       92            92            []
29722   1m-ovYop9xL3HiKjHjwFGg       27            27            []
31023   1td1l51xHG3tdd2wYxBF3w       26            26            []
31561   1x2ioOV3ikLNjkjOkq8Akw       26            26            []
33300   26E5aviHRj7lLb4UwTiJvg       25            25            []
...                        ...      ...           ...           ...
653889  wxAPTI-Wo3a2tE7OhnUP9A       45            45            []
657106  xFFmQTr0srCsnXiupIXT2A       39            39  [2012, 2013]
657186  xFdK5lgUdeT3bWC7_MvcyQ       25            25            []
660634  x_-VYseN-2Brf2EEzGw_Dw       53            53            []
661874  xgUC_evaln1piMg5oCWKrQ       21            21            []
662842  xmTkHgJRVDOHLE4eDcheMw       22            22            []
662903  xmpNnfZ9ugzcTlxKx1I1Sg       23            23            []
663039  xncGcrvha6w4uwSX2tlRIg       32            32            []
664905  xyuqtuV71y2F3uD9z97jPw       40            40            []
665557  y1jPEGJAGJkgrAbBIijWLw       28            28            []
666103  y4x-YtRk74Cig86jodTyJQ       35            35            []
667166  yBH29hOu2xB-6dHhl88NPQ       42            42            []
668005  yGjsNhvM3ZHMWbCXqUeeCg       29            29            []
668507  yJboJ2i9ZFsGe_XiyO5OSQ       25            25            []
671377  y_FoL3sc9vdjwwzthw8qBA       21            21            []
674368  yr55p6-XwbU8IJIsLueVAQ       39            39            []
674436  yrXCDuHsWLt14ql-aLOCYg       33            33            []
675241  ywBLk8e8QBan9WPgTuYgWQ       21            21            []
675820  yz_OZjFAfLT3074vjUAePQ       24            24            []
678881  zGz5NI4sscO4_6YYH-qVtA       89            89            []
680592  zRETaVK0hfu_LVSbaksSOA       80            80            []
680774  zSQps4R-_w8YkjDtgIsfiw       28            28            []
681428  zWUQmRH9KDpTGQAQWkKTsA       20            20            []
683333  zgvq1Av35k-CeW1GmhxJnw       29            29            []
684894  zptl58gf928g4JOOP2N0-w       55            55        [2016]
685400  zsvR-RUVTYedUUKkVe7Bkw       31            31            []
685587  zu7JY4ITfnDjoCu9we9biA       22            22            []
685947  zw7

In [206]:
user[user.user_id == '-1Q1s_NMGjBLBULA8z_npg']

average_stars                                        compliments  \
3770            4.0  {'writer': 14, 'more': 7, 'cute': 12, 'cool': ...   

                   elite  fans  \
3770  [2005, 2006, 2007]    19   

                                                friends   name  review_count  \
3770  [rpOyqD_893cqmDAtJLbdog, i63u3SdbrLsP4FxiSKP0Z...  Leila             1   

      type                 user_id                                 votes  \
3770  user  -1Q1s_NMGjBLBULA8z_npg  {'funny': 1, 'cool': 4, 'useful': 4}   

     yelping_since  
3770       2005-06

In [209]:
review[review.user_id == '-1Q1s_NMGjBLBULA8z_npg'].text.all()

"Our new favorite place to stay in Vegas. Yeah, it's kinda far down the strip, but who cares? the accomodation is top notch, with spacious rooms that has floor to ceiling windows, giving you a complete view of the strip. Everything in the room is indulgent, from plush slippers to the fluffy comforter to the chocolates. When I was hung over and slumped over in the bathtub, I was served wonton soup on a silver platter while the server informed me that the Wynn does not use eggs more than 2 days old. That's not I care  how old those eggs are but it stil sounds impressive."

In [302]:
business.columns

Index(['attributes', 'business_id', 'categories', 'city', 'full_address',
       'hours', 'latitude', 'longitude', 'name', 'neighborhoods', 'open',
       'review_count', 'stars', 'state', 'type'],
      dtype='object')

In [289]:
#business.categories.apply(lambda x: ','.join(x))

In [288]:
#business.categories.apply(lambda x: ','.join(x)).str.cat(sep=',')

In [286]:
unique_categories = set([category for cats in list(business.categories) for category in cats])

In [287]:
len(unique_categories)

1017

In [306]:
business_cats = business.groupby('categories')

In [310]:
business_cats.groups

TypeError: unhashable type: 'list'

In [337]:
business2 = business.copy()

In [339]:
business2.categories = business2.categories.apply(lambda x: tuple(x))

In [341]:
business2

attributes  \
0      {'Take-out': True, 'Has TV': False, 'Noise Lev...   
1      {'Outdoor Seating': False, 'Good For Groups': ...   
2                                {'Good for Kids': True}   
3      {'Good For Groups': True, 'Has TV': True, 'Smo...   
4      {'Parking': {'valet': False, 'validated': Fals...   
5      {'Good For Groups': True, 'Has TV': True, 'Smo...   
6                         {'Accepts Credit Cards': True}   
7                                {'Good for Kids': True}   
8                                                     {}   
9                                                     {}   
10                                     {'Wi-Fi': 'free'}   
11                                                    {}   
12     {'Take-out': True, 'Good For Groups': True, 'P...   
13     {'Parking': {'valet': False, 'validated': Fals...   
14                                                    {}   
15                        {'Accepts Credit Cards': True}   
16                                                    {}   
17     {'Dogs Allowed': False, 'Good For Groups': Tru...   
18     {'Good For Groups': False, 'Has TV': True, 'Co...   
19     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
20     {'Price Range': 2, 'Wi-Fi': 'free', 'Accepts C...   
21     {'Take-out': True, 'Has TV': True, 'Noise Leve...   
22                                                    {}   
23                       {'Accepts Credit Cards': False}   
24     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
25                         {'By Appointment Only': True}   
26     {'Take-out': True, 'Has TV': True, 'Noise Leve...   
27                                                    {}   
28     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
29     {'Parking': {'valet': False, 'validated': Fals...   
...                                                  ...   
85871  {'Parking': {'valet': False, 'validated': Fals...   
85872  {'Parking': {'valet': False, 'validated': Fals...   
85873                                                 {}   
85874  {'Outdoor Seating': True, 'Take-out': True, 'G...   
85875                                                 {}   
85876                                                 {}   
85877                                                 {}   
85878                            {'Good for Kids': True}   
85879                                                 {}   
85880  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85881  {'Price Range': 3, 'Take-out': True, 'Has TV':...   
85882  {'Parking': {'valet': False, 'validated': Fals...   
85883                                                 {}   
85884  {'Price Range': 1, 'Take-out': True, 'Has TV':...   
85885  {'Outdoor Seating': False, 'Take-out': True, '...   
85886  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85887  {'Parking': {'valet': False, 'validated': Fals...   
85888                     {'Accepts Credit Cards': True}   
85889  {'Dogs Allowed': True, 'Take-out': True, 'Has ...   
85890  {'Price Range': 2, 'Take-out': True, 'Has TV':...   
85891                                                 {}   
85892                     {'Accepts Credit Cards': True}   
85893                                                 {}   
85894  {'Parking': {'valet': False, 'validated': Fals...   
85895                                                 {}   
85896                     {'Accepts Credit Cards': True}   
85897  {'Take-out': True, 'Has TV': True, 'Noise Leve...   
85898                                                 {}   
85899  {'Price Range': 2, 'Wi-Fi': 'free', 'Accepts C...   
85900                                                 {}   

                  business_id  \
0      5UmKMjUEUNdYWqANhGckJw   
1      UsFtqoBl7naz8AVUBZMjQQ   
2      cE27W9VPgO88Qxe4ol6y_g   
3      mVHrayjG3uZ_RLHkLj-AMg   
4      mYSpR_SLPgUVymYOvTQd_Q   
5      KayYbHCt-RkbGcPdGOThNg   
6      b12U9TFESStdy7CsTtcOeg   
7      Sktj1eHQFuVa-M4bgnEh8g   
8      3ZVKmuK2l7uXPE6lXY4Db

In [342]:
business2_cats = business2.groupby('categories')

In [344]:
len(business2_cats.groups)

11054

In [349]:
business2_cats.groups

dict_items([(('Active Life', 'Rafting/Kayaking', 'Sporting Goods', 'Shopping', 'Paddleboarding'), Int64Index([2074], dtype='int64')), (('Local Services', 'Screen Printing', 'Printing Services', 'Graphic Design', 'Professional Services'), Int64Index([2614, 82534], dtype='int64')), (('American (Traditional)', 'Southern', 'Barbeque', 'Restaurants'), Int64Index([40328, 83333], dtype='int64')), (('Vehicle Shipping', 'Automotive'), Int64Index([22810, 49344, 83624, 85413], dtype='int64')), (('Mass Media', 'Marketing', 'Radio Stations', 'Advertising', 'Professional Services'), Int64Index([78740], dtype='int64')), (('Laboratory Testing', 'Medical Centers', 'Diagnostic Services', 'Health & Medical', 'Dialysis Clinics'), Int64Index([7547], dtype='int64')), (('Professional Services', 'Graphic Design', 'Vehicle Wraps', 'Web Design', 'Automotive'), Int64Index([71524], dtype='int64')), (('Bars', 'Breakfast & Brunch', 'Pubs', 'Nightlife', 'Canadian (New)', 'Restaurants'), Int64Index([53806], dtype='in

In [352]:
business.columns

Index(['attributes', 'business_id', 'categories', 'city', 'full_address',
       'hours', 'latitude', 'longitude', 'name', 'neighborhoods', 'open',
       'review_count', 'stars', 'state', 'type'],
      dtype='object')

In [370]:
business.attributes[['Pizza' in business.categories]]

C:\Users\ali_a\Anaconda3\lib\site-packages\pandas\core\internals.py:252: VisibleDeprecationWarning: boolean index did not match indexed array along dimension 0; dimension is 85901 but corresponding boolean dimension is 1
  return self.values[slicer]
C:\Users\ali_a\Anaconda3\lib\site-packages\pandas\indexes\base.py:1434: VisibleDeprecationWarning: boolean index did not match indexed array along dimension 0; dimension is 85901 but corresponding boolean dimension is 1
  result = getitem(key)


Series([], Name: attributes, dtype: object)

In [365]:
business.attributes.apply(lambda x: x.keys())

0        (Take-out, Has TV, Noise Level, Good For Group...
1        (Outdoor Seating, Good For Groups, Price Range...
2                                          (Good for Kids)
3        (Good For Groups, Has TV, Smoking, Attire, Acc...
4        (Parking, Good For Groups, Accepts Credit Card...
5        (Good For Groups, Has TV, Smoking, Attire, Acc...
6                                   (Accepts Credit Cards)
7                                          (Good for Kids)
8                                                       ()
9                                                       ()
10                                                 (Wi-Fi)
11                                                      ()
12       (Take-out, Good For Groups, Price Range, Attir...
13       (Parking, Take-out, Price Range, Accepts Credi...
14                                                      ()
15                                  (Accepts Credit Cards)
16                                                      

# Building a recommender system
First, get the relevant data: user-business stars... create matrices Y, R

** Y ** that contains #business as rows, #user as columns, values = stars

** R ** same size as Y, values = 1 if a user has rated a business **



In [373]:
# Need to determine businesses and users to include in the model

In [377]:
review.head(1)

business_id       date               review_id  stars  \
0  5UmKMjUEUNdYWqANhGckJw 2012-08-01  Ya85v4eqdd6k9Od8HbQjyA      4   

                                                text    type  \
0  Mr Hoagie is an institution. Walking in, it do...  review   

                  user_id                                 votes  
0  PUFPaY9KxDAcGqfsorJp3Q  {'funny': 0, 'cool': 0, 'useful': 0}

In [379]:
len(review.business_id.unique())

85539

In [380]:
len(review.user_id.unique())

686556

In [382]:
np.zeros([85539,686556])

MemoryError: 

In [383]:
# city-based grouping and recommendation

In [385]:
rev_biz = review.merge(business, right_on='business_id', left_on='business_id')

In [386]:
rev_biz.columns

Index(['business_id', 'date', 'review_id', 'stars_x', 'text', 'type_x',
       'user_id', 'votes', 'attributes', 'categories', 'city', 'full_address',
       'hours', 'latitude', 'longitude', 'name', 'neighborhoods', 'open',
       'review_count', 'stars_y', 'state', 'type_y'],
      dtype='object')

In [390]:
rev_biz.shape

(2685066, 22)

In [388]:
rev_biz_usr = rev_biz.merge(user, left_on = 'user_id', right_on='user_id')

In [389]:
rev_biz_usr.shape

(2685066, 32)

In [410]:
col_drop = rev_biz_usr.columns.difference(['business_id', 'user_id', 'stars_x'])
col_drop

Index(['attributes', 'average_stars', 'categories', 'city', 'compliments',
       'date', 'elite', 'fans', 'friends', 'full_address', 'hours', 'latitude',
       'longitude', 'name_x', 'name_y', 'neighborhoods', 'open',
       'review_count_x', 'review_count_y', 'review_id', 'stars_y', 'state',
       'text', 'type', 'type_x', 'type_y', 'votes_x', 'votes_y',
       'yelping_since'],
      dtype='object')

In [412]:
rev_biz_usr.drop(col_drop, axis=1, inplace=True)

In [413]:
rev_biz_usr.columns

Index(['business_id', 'stars_x', 'user_id'], dtype='object')

In [767]:
len(business.business_id.unique())

85900

In [768]:
len(user.user_id.unique())

686556

## The following pivot table gives an error due to out of memory, not negagtive dimensions:
memory needed assuming 8 bytes per value is:
686556 x 85900 x 8 / 1024 / 1024/ 1024 = 440 GB

In [769]:
686556 * 85900 * 8 / 1024 / 1024/ 1024

439.3991857767105

In [423]:
rev_biz_usr.pivot(index='business_id', columns='user_id', values='stars_x')

ValueError: negative dimensions are not allowed

In [437]:
Y = pd.pivot_table(rev_biz_usr, index='business_id', columns='user_id', values='stars_x', dropna=False, aggfunc='mean')

ValueError: negative dimensions are not allowed

In [418]:
df = pd.DataFrame({'foo': ['one','one','two','two','two'],
                       'bar': ['A', 'B', 'A', 'B', 'C'],
                       'baz': [1, 2, 4, 5, 6]})

In [419]:
df

bar  baz  foo
0   A    1  one
1   B    2  one
2   A    4  two
3   B    5  two
4   C    6  two

In [421]:
df.pivot(index='bar', columns='foo', values='baz')

foo  one  two
bar          
A    1.0  4.0
B    2.0  5.0
C    NaN  6.0

In [436]:
# !pip install geopy

In [439]:
from geopy.distance import vincenty

In [465]:
vincenty((41.49008, -71.312796), (41.499498, -81.695391)).km

866.4554329011002

In [469]:
vincenty((89, -71.312796), (89, -81.695391)).km

20.211398233242566

In [482]:
tuple(business.iloc[0][['latitude','longitude']])

(40.3543266, -79.900705700000003)

In [487]:
tuple(business.iloc[1][['latitude','longitude']])

(40.350552700000001, -79.886813799999999)

In [771]:
import shapely

In [772]:
shapely.__version__

'1.5.17'

In [ ]:
!conda install -c conda-forge geopandas

In [ ]:
import geopandas

# assume 30 km as maximum distance willing to travel, to find vicinity within state

In [639]:
business.groupby(['state']).filter(lambda r: len(r) < 20).state.unique()

array(['CA', 'FL', 'NM', 'TX', 'HAM', 'SCB', 'ELN', 'FIF', 'NTH', 'XGL',
       'RP', 'KHL', 'MN', 'NW', 'TAM', 'AL', 'AK'], dtype=object)

In [640]:
# ignore states with less than 20 businesses... we have 12 states with more than 20 businesses
business_state = business.groupby(['state']).filter(lambda r: len(r) >= 20)

In [638]:
business_state.groupby('state').size()

state
AZ     36500
BW      1055
EDH     3297
IL       808
MLN      161
NC      6835
NV     23591
ON       530
PA      4086
QC      5591
SC       325
WI      3066
dtype: int64

In [486]:
vincenty(tuple(business.iloc[0][['latitude','longitude']]), tuple(business.iloc[1][['latitude','longitude']])).km

1.252354521138835

In [513]:
business.business_id[1]

'UsFtqoBl7naz8AVUBZMjQQ'

In [520]:
nearby ={}
nearby[business.business_id[0]]=[]
nearby[business.business_id[0]].extend([business.business_id[1]])
nearby[business.business_id[0]].extend([business.business_id[190]])
nearby[business.business_id[0]].extend([business.business_id[2500]])
nearby

{'5UmKMjUEUNdYWqANhGckJw': ['UsFtqoBl7naz8AVUBZMjQQ',
  'bAnFMk-UoNjyVYUj0sJT0w',
  'CAD0tKIttOTTgT7TMJa-lQ']}

In [740]:
nearby = {}
nearby['a'] = [4]
nearby

{'a': [4]}

In [741]:
try:
    nearby['b'].extend(['a'])
except KeyError as e:
    nearby['b'] = ['a']

In [742]:
nearby

{'a': [4], 'b': ['a']}

In [735]:
nearby['b'] = nearby.get('b', []) + [5]

In [728]:
nearby['b'] = nearby.get('b', [])

In [727]:
nearby['b'].extend(['a'])

In [736]:
nearby

{'a': [4], 'b': ['a', 5, 5]}

In [726]:
nearby.get('b', [])

[8, 8, 'a', 'a', 'a', 'a']

In [760]:
group = business_groups.get_group('AZ')

attributes  \
2288   {'Parking': {'valet': False, 'validated': Fals...   
2376    {'Price Range': 2, 'Accepts Credit Cards': True}   
2377   {'Take-out': True, 'Has TV': False, 'Noise Lev...   
2378   {'Price Range': 1, 'Take-out': True, 'Noise Le...   
2379                                                  {}   
2380                                                  {}   
2381                                                  {}   
2382   {'Dogs Allowed': False, 'Good For Groups': Tru...   
2383   {'Outdoor Seating': True, 'Has TV': False, 'No...   
2384                                                  {}   
2385                      {'Accepts Credit Cards': True}   
2386   {'Take-out': True, 'Has TV': True, 'Noise Leve...   
2387   {'Price Range': 2, 'Wheelchair Accessible': Tr...   
2388   {'Parking': {'valet': False, 'validated': Fals...   
2389                       {'By Appointment Only': True}   
2390   {'Parking': {'valet': False, 'validated': True...   
2391    {'Price Range': 2, 'Accepts Credit Cards': True}   
2392   {'Price Range': 1, 'Wi-Fi': 'free', 'Accepts C...   
2393   {'Parking': {'valet': False, 'validated': Fals...   
2394   {'Attire': 'casual', 'Take-out': True, 'Has TV...   
2395   {'Take-out': True, 'Good For Groups': True, 'P...   
2396   {'Price Range': 3, 'Accepts Credit Cards': Tru...   
2397   {'Take-out': True, 'Good For Groups': False, '...   
2398   {'Take-out': True, 'Has TV': False, 'Good For ...   
2399                                                  {}   
2400   {'Parking': {'valet': False, 'validated': Fals...   
2401                             {'Good for Kids': True}   
2402   {'Dogs Allowed': False, 'Parking': {'valet': F...   
2403                                                  {}   
2404   {'Take-out': True, 'Has TV': False, 'Noise Lev...   
...                                                  ...   
85829  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85830                           {'Good for Kids': False}   
85831  {'Take-out': True, 'Has TV': False, 'Good For ...   
85835  {'Parking': {'valet': False, 'validated': Fals...   
85838                                                 {}   
85842  {'Parking': {'valet': False, 'validated': Fals...   
85846  {'Parking': {'valet': False, 'validated': Fals...   
85848                                                 {}   
85849  {'Accepts Credit Cards': True, 'By Appointment...   
85850  {'Take-out': True, 'Accepts Credit Cards': Tru...   
85851  {'Parking': {'valet': False, 'validated': Fals...   
85854                                                 {}   
85858                                                 {}   
85859  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85861                                                 {}   
85866  {'Accepts Credit Cards': True, 'Good for Kids'...   
85870  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85873                                                 {}   
85874  {'Outdoor Seating': True, 'Take-out': True, 'G...   
85880  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85881  {'Price Range': 3, 'Take-out': True, 'Has TV':...   
85882  {'Parking': {'valet': False, 'validated': Fals...   
85884  {'Price Range': 1, 'Take-out': True, 'Has TV':...   
85887  {'Parking': {'valet': False, 'validated': Fals...   
85889  {'Dogs Allowed': True, 'Take-out': True, 'Has ...   
85890  {'Price Range': 2, 'Take-out': True, 'Has TV':...   
85891                                                 {}   
85894  {'Parking': {'valet': False, 'validated': Fals...   
85897  {'Take-out': True, 'Has TV': True, 'Noise Leve...   
85898                                                 {}   

                  business_id  \
2288   LUvxyWag5gpiubSxOvR92A   
2376   Y5t01ll9I9GtSCtqHVv2DQ   
2377   x5Mv61CnZLohZWxfCVCPTQ   
2378   2ZnCITVa0abGce4gZ6RhIw   
2379   fmieXX46NLuR_0sJ-UWIHA   
2380   CCpYuGX2EXXbwJIDAsJDJw   
2381   gO2stAxnesXIejH6TMfD4g   
2382   EmzaQR5hQlF0WIl24NxAZA   
2383   SiwN7f0N4bs4ZtPc4yPgi

In [ ]:
if group.latitude.iloc[0] < 80:
    pass

In [766]:
nearby ={}
max_nearbies = 0

business_groups = business_state.groupby('state')

for name, group in business_groups:
    for i in range(0, len(group)):
        if group.latitude.iloc[i] < 80:
            near_group = group[(group.latitude < (group.latitude.iloc[i] + 1)) & (group.latitude > (group.latitude.iloc[i] - 1)) & 
                          (group.longitude < (group.longitude.iloc[i] + 1)) & (group.longitude > (group.longitude.iloc[i] - 1))]
        for j in range(0, len(near_group)):

            distance = vincenty(tuple(group.iloc[i][['latitude','longitude']]), tuple(near_group.iloc[j][['latitude','longitude']])).km
            if distance < 30:
                c += 1
                try:
                    nearby[group.business_id.iloc[i]].extend([near_group.business_id.iloc[j]])
                except KeyError as e:
                    nearby[group.business_id.iloc[i]] = [near_group.business_id.iloc[j]]
                try:
                    nearby[near_group.business_id.iloc[j]].extend([group.business_id.iloc[i]])
                except KeyError as e:
                    nearby[near_group.business_id.iloc[j]] = [group.business_id.iloc[i]]
        if max_nearbies < c:
            max_nearbies = c
    print (len(group))


KeyboardInterrupt: 

In [755]:
business_state[(abs(business_state.latitude) < (business_state.latitude[0] + 1))]

attributes  \
0      {'Take-out': True, 'Has TV': False, 'Noise Lev...   
1      {'Outdoor Seating': False, 'Good For Groups': ...   
2                                {'Good for Kids': True}   
3      {'Good For Groups': True, 'Has TV': True, 'Smo...   
4      {'Parking': {'valet': False, 'validated': Fals...   
5      {'Good For Groups': True, 'Has TV': True, 'Smo...   
6                         {'Accepts Credit Cards': True}   
7                                {'Good for Kids': True}   
8                                                     {}   
9                                                     {}   
10                                     {'Wi-Fi': 'free'}   
11                                                    {}   
12     {'Take-out': True, 'Good For Groups': True, 'P...   
13     {'Parking': {'valet': False, 'validated': Fals...   
14                                                    {}   
15                        {'Accepts Credit Cards': True}   
16                                                    {}   
17     {'Dogs Allowed': False, 'Good For Groups': Tru...   
18     {'Good For Groups': False, 'Has TV': True, 'Co...   
19     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
20     {'Price Range': 2, 'Wi-Fi': 'free', 'Accepts C...   
21     {'Take-out': True, 'Has TV': True, 'Noise Leve...   
22                                                    {}   
23                       {'Accepts Credit Cards': False}   
24     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
25                         {'By Appointment Only': True}   
26     {'Take-out': True, 'Has TV': True, 'Noise Leve...   
27                                                    {}   
28     {'Take-out': True, 'Has TV': False, 'Noise Lev...   
29     {'Parking': {'valet': False, 'validated': Fals...   
...                                                  ...   
85866  {'Accepts Credit Cards': True, 'Good for Kids'...   
85867  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85868  {'Parking': {'valet': False, 'validated': Fals...   
85869  {'Outdoor Seating': True, 'Has TV': True, 'Whe...   
85870  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85871  {'Parking': {'valet': False, 'validated': Fals...   
85873                                                 {}   
85874  {'Outdoor Seating': True, 'Take-out': True, 'G...   
85875                                                 {}   
85878                            {'Good for Kids': True}   
85879                                                 {}   
85880  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85881  {'Price Range': 3, 'Take-out': True, 'Has TV':...   
85882  {'Parking': {'valet': False, 'validated': Fals...   
85883                                                 {}   
85884  {'Price Range': 1, 'Take-out': True, 'Has TV':...   
85885  {'Outdoor Seating': False, 'Take-out': True, '...   
85886  {'Take-out': True, 'Has TV': False, 'Noise Lev...   
85887  {'Parking': {'valet': False, 'validated': Fals...   
85888                     {'Accepts Credit Cards': True}   
85889  {'Dogs Allowed': True, 'Take-out': True, 'Has ...   
85890  {'Price Range': 2, 'Take-out': True, 'Has TV':...   
85891                                                 {}   
85892                     {'Accepts Credit Cards': True}   
85894  {'Parking': {'valet': False, 'validated': Fals...   
85895                                                 {}   
85896                     {'Accepts Credit Cards': True}   
85897  {'Take-out': True, 'Has TV': True, 'Noise Leve...   
85898                                                 {}   
85900                                                 {}   

                  business_id  \
0      5UmKMjUEUNdYWqANhGckJw   
1      UsFtqoBl7naz8AVUBZMjQQ   
2      cE27W9VPgO88Qxe4ol6y_g   
3      mVHrayjG3uZ_RLHkLj-AMg   
4      mYSpR_SLPgUVymYOvTQd_Q   
5      KayYbHCt-RkbGcPdGOThNg   
6      b12U9TFESStdy7CsTtcOeg   
7      Sktj1eHQFuVa-M4bgnEh8g   
8      3ZVKmuK2l7uXPE6lXY4Db

In [ ]:
len(nearby)

In [ ]:
import pickle
# save the nearby dictionary as it take a long time to build
with open('nearby_businesses_dict.pickle', 'wb') as handle:
    pickle.dump(nearby, handle, protocol=pickle.HIGHEST_PROTOCOL)
# To load the dictionary un-comment the following:___
# with open('nearby_businesses_dict.pickle', 'rb') as handle:
#     b = pickle.load(handle)

In [534]:
# nearby ={}
# max_nearbies = 0
# business_groups = business_state.groupby('state')
# for group in business_groups:
    
    
    
# for i in range(0, len(business)):
#     nearby[business.business_id.iloc[i]]=[]
#     c = 0
#     for j in range(0, len(business)):
#         if i==j: continue
#         distance = vincenty(tuple(business.iloc[i][['latitude','longitude']]), tuple(business.iloc[j][['latitude','longitude']])).km
#         if distance < 30:
#             c += 1
#             nearby[business.business_id.iloc[i]].extend([business.business_id.iloc[j]])        
#     if c > max_nearbies:
#         max_nearbies = c

In [550]:
len(nearby[business.business_id.iloc[250]])

29

In [471]:
len(business)

85901

In [463]:
85901*85901

7378981801

In [554]:
b = set([1,2,3])

In [613]:
set(business.state[(business.longitude < 0)].values).intersection({1,2,'NY'})

set()

In [612]:
(business.state.unique())

array(['PA', 'NC', 'SC', 'WI', 'IL', 'AZ', 'CA', 'NV', 'FL', 'NM', 'QC',
       'ON', 'TX', 'EDH', 'MLN', 'HAM', 'SCB', 'ELN', 'FIF', 'NTH', 'XGL',
       'BW', 'RP', 'KHL', 'MN', 'NW', 'TAM', 'AL', 'AK'], dtype=object)

In [611]:
business[business.state == '']

Empty DataFrame
Columns: [attributes, business_id, categories, city, full_address, hours, latitude, longitude, name, neighborhoods, open, review_count, stars, state, type]
Index: []

In [610]:
set(business.state[(business.longitude < 0) & (business.latitude < 50)].values)

{'AK',
 'AL',
 'AZ',
 'CA',
 'FL',
 'IL',
 'MN',
 'NC',
 'NM',
 'NV',
 'ON',
 'PA',
 'QC',
 'SC',
 'TAM',
 'TX',
 'WI'}

In [563]:
business.iloc[47394]

attributes                           {'Price Range': 3}
business_id                      W2RFJg1N_MzuI8I42MQ2VA
categories       [Department Stores, Fashion, Shopping]
city                                          Karlsruhe
full_address            Windeckstr. 15\n76135 Karlsruhe
hours                                                {}
latitude                                        48.9881
longitude                                       8.37332
name                                     Heinrich Heine
neighborhoods                                        []
open                                               True
review_count                                          5
stars                                                 3
state                                                BW
type                                           business
Name: 47394, dtype: object

In [ ]:
if __name__ == '__main__':
    main() 